### PR TITLE
fix: limit concurrent database connections in prebuild reconciliation

### LINKED ISSUE #20606: Limit number of concurrent database connections used by prebuilds
## Problem 

The prebuilds reconciliation loop uses an advisory lock to ensure only one replica executes reconciliation at a time. During reconciliation, we spawn one goroutine per preset to calculate and execute reconciliation actions.

While most goroutines result in 0 reconciliation actions (no database operations), we can still end up with many concurrent database write operations (creates/deletes of prebuilds). This means we can easily exceed the maximum number of open concurrent database connections per replica ([10 connections](https://github.com/coder/coder/blob/main/cli/server.go#L2401)). This can exhaust the database connection pool, causing both these goroutines and other parts of the system (such as API endpoints) to wait for available connections.

### 1) Filter presets earlier (optimization)

Update `GetTemplatePresetsWithPrebuilds` to filter out inactive/deleted template versions at the database level. Only include inactive/deleted versions if they have prebuilds that still need reconciliation (e.g., running workspaces that need to be deleted). This would significantly reduce the number of presets returned and consequently the number of goroutines spawned. 

For reference, a preset in this context is a tuple `(organization, template, version, preset name)`. On dogfood we currently have 3121 presets returned by `GetTemplatePresetsWithPrebuilds`, the majority of which correspond to old presets that no longer need reconciliation. Once a version becomes inactive, in a normal scenario, it should take 1 reconciliation cycle to clean up the prebuilds for that version.

### 2) Limit concurrent write operations

Limit the number of concurrent write operations to 2-3 connections for reconciliation to avoid impacting other parts of the system that require database connections.

**Option A: Use errgroup concurrency limiting**

Add `eg.SetLimit(5)` to limit concurrent goroutines. This is simple (one line of code) and allows reasonable parallelism, but limits all goroutines, even those that don't need database connections, most goroutines with no actions will hold a slot but do no database operation. However, this impact should be significantly reduced by the filtering optimization above.

**Option B: Separate worker pool for writes**

Keep goroutines for calculating actions (fast, in-memory), but funnel actual write operations through a dedicated separate worker pool of 2-3 workers. This provides maximum parallelism for the calculation phase and precise control over database connection usage, but adds complexity with worker pool coordination.

### PR DESCRIPTION
**Pull request by ssncferreira**
## Description

This PR addresses database connection pool exhaustion during prebuilds reconciliation by introducing two changes:
* `CanSkipReconciliation`: Filters out presets that don't need reconciliation before spawning goroutines. This ensures we only create goroutines for presets that will (_most likely_) perform database operations, avoiding unnecessary connection pool usage.
* Dynamic `eg.SetLimit`: Limits concurrent goroutines based on the configured database connection pool size (`CODER_PG_CONN_MAX_OPEN / 2`). This replaces the previous hardcoded limit of 5, ensuring the reconciliation loop scales appropriately with the configured pool size while leaving capacity for other database operations.

## Changes

* Add `CanSkipReconciliation()` method to `PresetSnapshot` that returns true for inactive presets with no running workspaces, no pending jobs, or expired prebuilds.
* Add `maxDBConnections` parameter to `NewStoreReconciler` and compute `reconciliationConcurrency` as half the pool size (minimum 1).
* Add `ReconciliationConcurrency()` getter method to `StoreReconciler`.
* Add `eg.SetLimit(c.reconciliationConcurrency)` to bound concurrent reconciliation goroutines.
* Add `PresetsTotal` and `PresetsReconciled` to `ReconcileStats` for observability.
* Add `TestCanSkipReconciliation` unit tests.
* Add `TestReconciliationConcurrency` unit tests.
* Add benchmark tests for reconciliation performance.

## Benchmarks

* `BenchmarkReconcileAll_NoOps`: Tests presets with no reconciliation actions. All presets are filtered by `CanSkipReconciliation`, resulting in no goroutines spawned and no database connections used.
* `BenchmarkReconcileAll_ConnectionContention`: Tests presets where all require reconciliation actions. All presets spawn goroutines, but concurrency is limited by `eg.SetLimit(reconciliationConcurrency)`.
* `BenchmarkReconcileAll_Mix`: Simulates a realistic scenario with a large subset of inactive presets (filtered by `CanSkipReconciliation`) and a smaller subset requiring reconciliation (limited by `eg.SetLimit`).

Closes: https://github.com/coder/coder/issues/20606

### CODE REVIEW

**Comment by ssncferreira** [comment]
## Benchmarks

<details>
<summary>Setup</summary>

Benchmark tests were conducted in a dogfood workspace:
```
goos: linux
goarch: amd64
pkg: github.com/coder/coder/v2/enterprise/coderd/prebuilds
cpu: AMD EPYC 9454P 48-Core Processor
```

Using the Go test tool with the following configurations:
```
GOMAXPROCS=16
-benchtime=5x
-count=2
```

The baseline benchmark was taken on the `main` branch at commit https://github.com/coder/coder/commit/39bf9ed18aaaa785cbb5fad39f8a1ed704354cca.

To analyze performance differences between the baseline (main) and the proposed changes, [benchstat](https://pkg.go.dev/golang.org/x/perf/cmd/benchstat) was used to compare the benchmark outputs.

</details>

### Benchmark Results

The table below shows the geometric mean of reconciliation loop duration for each benchmark. Full results are attached.

| Benchmark | main | PR | Change |
|:--------:|:--------:|:--------:|:--------:|
| NoOps | 53.34ms | 49.69ms | -6.84% |
| ConnectionContention | 597.9ms | 904.3ms | +51.23% |
| Mix | 167.2ms | 190.2ms | +13.76% |

<details>
<summary>Raw benchmark files</summary>

**BenchmarkReconcileAll_NoOps**

[main_results](https://github.com/user-attachments/files/24392852/main_noop.txt)
[pr_results](https://github.com/user-attachments/files/24392855/pr_20908_noop.txt)
[benchstat](https://github.com/user-attachments/files/24392856/benchstat_noop.txt)

**BenchmarkReconcileAll_ConnectionContention**

[main_results](https://github.com/user-attachments/files/24392862/main_db_contention.txt)
[pr_results](https://github.com/user-attachments/files/24392864/pr_20908_db_contention.txt)
[benchstat](https://github.com/user-attachments/files/24392866/benchstat_db_contention.txt)

**BenchmarkReconcileAll_Mix**

[main_results](https://github.com/user-attachments/files/24392868/main_mix.txt)
[pr_results](https://github.com/user-attachments/files/24392870/pr_20908_mix.txt)
[benchstat](https://github.com/user-attachments/files/24392871/benchstat_mix.txt)

</details>

### Benchmark Analysis

* `NoOps` (-6.84%): Slight improvement when all presets are inactive. `CanSkipReconciliation` filters out all presets before goroutine creation, avoiding unnecessary overhead.
* `ConnectionContention` (+51.23%): Expected slowdown when all presets require database operations. `eg.SetLimit(5)` throttles concurrency to protect the connection pool. This is the tradeoff from this solution: slower reconciliation in exchange for bounded resource usage that won't starve other system components.
* `Mix` (+13.76%): Moderate slowdown in realistic scenarios. Most presets are filtered by `CanSkipReconciliation`, while the smaller active subset is processed with limited concurrency.

File: coderd/prebuilds/preset_snapshot.go [hunk 1]
```diff
@@ -85,6 +85,39 @@ func NewPresetSnapshot(
...
+// in-progress builds.
+//
+// Fields checked:
+//   - isActive(): Active presets are never skipped. Presets from active template
+//     versions always go through the reconciliation loop to ensure desired_instances
+//     is maintained correctly.
+//   - Running: Non-empty means there are prebuilds to delete.
+//   - Expired: Non-empty means there are expired prebuilds to delete.
+//   - PendingCount: Non-zero means there are pending jobs to cancel.
+//   - Backoff: Checked defensively, but only populated for active presets.
```
Comment (dannykopping) [inline coderd/prebuilds/preset_snapshot.go]: Nit: I'd prefer these comments to be inline to ensure they get updated if/when the logic changes. Coupling the comments to the actual lines will hopefully prevent one from changes without the other.

File: coderd/prebuilds/preset_snapshot.go [hunk 2]
```diff
@@ -85,6 +85,39 @@ func NewPresetSnapshot(
...
+func (p PresetSnapshot) CanSkipReconciliation() bool {
+	// Only skip presets from inactive template versions that have absolutely nothing to clean up
+	if !p.isActive() &&
+		len(p.Running) == 0 &&
+		len(p.Expired) == 0 &&
+		p.PendingCount == 0 &&
+		p.Backoff == nil {
+		return true
+	}
+	return false
```
Comment (dannykopping) [inline coderd/prebuilds/preset_snapshot.go]: I'd suggest splitting these up into individual branches. Right now it'll be impossible to assess the coverage of this function, and since it's such a crucial function it needs to be 100% covered.

```suggestion
    if p.IsActive() { return false }
    if len(p.Running) > 0 { return false }
    ... etc ...
    
    return true
```

File: enterprise/coderd/prebuilds/reconcile.go [hunk 3]
```diff
@@ -341,6 +345,15 @@ func (c *StoreReconciler) ReconcileAll(ctx context.Context) (stats prebuilds.Rec
 		}
 
 		var eg errgroup.Group
+		// Limit concurrent goroutines to avoid exhausting the database connection pool.
+		// Each preset reconciliation may perform multiple sequential database operations
+		// (creates/deletes), and with a pool limit of 10 connections, allowing unlimited
+		// concurrency would cause connection contention and thrashing. A limit of 5 ensures
+		// we stay below the pool limit while maintaining reasonable parallelism.
+		eg.SetLimit(5)
```
Comment (dannykopping) [inline enterprise/coderd/prebuilds/reconcile.go]: We'll [soon](https://github.com/coder/coder/pull/21403) support configurable connection pool sizes. I think this should be a function of the pool size (i.e. always max half). This PR may need to wait for the linked PR to land before we can do this, or we can add a follow-up to that PR to enforce this arithmetic.

File: enterprise/coderd/prebuilds/reconcile.go [hunk 4]
```diff
@@ -349,6 +362,17 @@ func (c *StoreReconciler) ReconcileAll(ctx context.Context) (stats prebuilds.Rec
 				continue
 			}
 
+			// Performance optimization: Skip presets that won't need any database operations.
+			// Presets from inactive template versions with no running workspaces, pending jobs,
+			// or in-progress builds will result in no reconciliation actions (no creates/deletes).
```
Comment (dannykopping) [inline enterprise/coderd/prebuilds/reconcile.go]: A piece of advice: comments become stale the moment they're written. The reader can reference the method's comments to get full context.


```suggestion
```

**Review by dannykopping — COMMENTED** [review]
Haven't reviewed the tests/benchmarks yet, but wanted to give feedback on what I've read so far.

File: coderd/prebuilds/preset_snapshot.go (hunk 1 above)
Comment (ssncferreira) [reply coderd/prebuilds/preset_snapshot.go]: Addressed in https://github.com/coder/coder/pull/20908/commits/e4490529bdbe545d34d065d6829dc54efb49752e

File: coderd/prebuilds/preset_snapshot.go (hunk 2 above)
Comment (ssncferreira) [reply coderd/prebuilds/preset_snapshot.go]: That is a good point, and this is cleaner. Addressed in https://github.com/coder/coder/pull/20908/commits/e4490529bdbe545d34d065d6829dc54efb49752e

File: enterprise/coderd/prebuilds/reconcile.go (hunk 4 above)
Comment (ssncferreira) [reply enterprise/coderd/prebuilds/reconcile.go]: Addressed in https://github.com/coder/coder/pull/20908/commits/e4490529bdbe545d34d065d6829dc54efb49752e

File: coderd/prebuilds/preset_snapshot.go [hunk 5]
```diff
@@ -85,6 +85,49 @@ func NewPresetSnapshot(
...
+	}
+
+	// Inactive presets with running prebuilds means there are prebuilds to delete.
+	if len(p.Running) > 0 {
+		return false
+	}
+
+	// Inactive presets with expired prebuilds means there are expired prebuilds to delete.
+	if len(p.Expired) > 0 {
+		return false
```
Comment (dannykopping) [inline coderd/prebuilds/preset_snapshot.go:106]: Thanks for the refactor 👍 
This case is missing coverage.

Comment (ssncferreira) [reply coderd/prebuilds/preset_snapshot.go:106]: As discussed previously, this is a bug in the reconciliation logic where expired prebuilds from inactive versions are not deleted. I will address this in a separate PR and then add coverage for it in this test.

File: enterprise/coderd/prebuilds/reconcile.go (hunk 3 above)
Comment (SasSwart) [reply enterprise/coderd/prebuilds/reconcile.go]: I'm tending to the other PR. I think we can merge this PR and address it in the other.

**Review by SasSwart — APPROVED** [review]
Solid work, thanks!

Some small thoughts:
* Our benchmarks currently assume the same connection pool size as we hardcode right now. This assumption will soon no longer hold as we allow a configurable connection pool size. I spent some time thinking about whether pool size would influence these benchmark tests, but I don't think it will. A larger pool size will make contention less likely. But A larger pool size that is still oversubscribed should exhibit the same symptoms we see for a smaller pool that is oversubscribed.
* We now have tracing for prebuilds. It occurs to me that we could consider wiring up metrics, tracing and profiling for benchmarks to be able to present their findings graphically in PRs. I'm not asking this of you right now, its just a thought that I had and might play with.

**Comment by dannykopping** [comment]
https://github.com/coder/coder/pull/21403 has landed, we can proceed with this now.

File: enterprise/coderd/prebuilds/reconcile.go (hunk 3 above)
Comment (ssncferreira) [reply enterprise/coderd/prebuilds/reconcile.go]: Addressed in https://github.com/coder/coder/pull/20908/commits/4cabf2de2c70a90a5a6b6645658e751e79a912c8

**Comment by ssncferreira** [comment]
> #21403 has landed, we can proceed with this now.

Addressed in https://github.com/coder/coder/pull/20908/commits/4cabf2de2c70a90a5a6b6645658e751e79a912c8

File: enterprise/coderd/prebuilds/reconcile.go [hunk 6]
```diff
@@ -93,20 +95,40 @@ func NewStoreReconciler(store database.Store,
 	notifEnq notifications.Enqueuer,
 	buildUsageChecker *atomic.Pointer[wsbuilder.UsageChecker],
 	tracerProvider trace.TracerProvider,
+	maxDBConnections int,
 ) *StoreReconciler {
+	// Limit concurrent preset reconciliation goroutines to avoid exhausting the
+	// coderd database connection pool. Each preset may perform multiple sequential
+	// database operations (creates/deletes). Using half of the pool size leaves
+	// capacity for other database operations while maintaining reasonable parallelism.
+	reconciliationConcurrency := maxDBConnections / 2
```
Comment (dannykopping) [inline enterprise/coderd/prebuilds/reconcile.go]: [The new default will be 30](https://github.com/coder/coder/pull/21572), so I worry that half may be too many.

What about something like `if min(maxDBConnections, 5) < 5 then maxDBConnections / 2 else 5`?

File: enterprise/coderd/prebuilds/reconcile.go [hunk 7]
```diff
@@ -93,20 +95,40 @@ func NewStoreReconciler(store database.Store,
...
+	if reconciliationConcurrency < 1 {
+		reconciliationConcurrency = 1
+	}
+
+	if maxDBConnections <= 0 {
+		logger.Error(context.Background(), "maxDBConnections is not positive, defaulting reconciliation concurrency to 1",
+			slog.F("max_db_connections", maxDBConnections))
+	}
```
Comment (dannykopping) [inline enterprise/coderd/prebuilds/reconcile.go]: This value is already validated to be `>= 0` on startup.

File: enterprise/coderd/coderd.go [hunk 8]
```diff
@@ -1311,7 +1311,18 @@ func (api *API) setupPrebuilds(featureEnabled bool) (agplprebuilds.Reconciliatio
...
+		api.Pubsub,
+		api.AGPL.FileCache,
+		api.DeploymentValues.Prebuilds,
+		api.Logger.Named("prebuilds"),
+		quartz.NewReal(),
+		api.PrometheusRegistry,
+		api.NotificationsEnqueuer,
+		api.AGPL.BuildUsageChecker,
+		api.TracerProvider,
+		int(api.DeploymentValues.PostgresConnMaxOpen.Value()),
```
Comment (dannykopping) [inline enterprise/coderd/coderd.go:1325]: Nit: I wonder if we should just pass `api.DeploymentValues` in, since we're now referencing it twice.

File: enterprise/coderd/prebuilds/reconcile.go (hunk 6 above)
Comment (ssncferreira) [reply enterprise/coderd/prebuilds/reconcile.go]: That's a good point 👀  Capping at 5 makes sense for now, though this might not scale well for large deployments that want more parallelism. If that becomes an issue, we might need to add a configuration option to specify the concurrency level (with validation to ensure it stays below the configured pool size).

File: enterprise/coderd/coderd.go (hunk 8 above)
Comment (ssncferreira) [reply enterprise/coderd/coderd.go:1325]: We could, but passing all of `DeploymentValues` would expose a lot more than the reconciler actually needs. I think this approach is clean in the sense that we are specific to what is being passed.

Comment (dannykopping) [reply enterprise/coderd/coderd.go:1325]: Cool, feel free to ignore 👍

File: enterprise/coderd/prebuilds/reconcile.go (hunk 6 above)
Comment (dannykopping) [reply enterprise/coderd/prebuilds/reconcile.go]: Sounds good 👍

Comment (ssncferreira) [reply enterprise/coderd/prebuilds/reconcile.go]: Addressed in https://github.com/coder/coder/pull/20908/commits/196192d4e114b0947d75624444dd2365a31c3dc8

File: enterprise/coderd/prebuilds/reconcile.go [hunk 9]
```diff
@@ -129,6 +139,33 @@ func NewStoreReconciler(store database.Store,
...
 }
 
+// calculateReconciliationConcurrency determines the number of concurrent
+// goroutines for preset reconciliation. Each preset may perform multiple
+// database operations (creates/deletes), so we limit concurrency to avoid
+// exhausting the connection pool while maintaining reasonable parallelism.
+//
+// Uses half the pool size, with a minimum of 1 and a maximum of 5.
+// TODO(ssncferreira): If this becomes a bottleneck, consider adding a configuration option.
+func calculateReconciliationConcurrency(maxDBConnections int) int {
```
Comment (dannykopping) [inline enterprise/coderd/prebuilds/reconcile.go:149]: Can we add some unit tests for this, just to be safe from future changes?

**Review by dannykopping — APPROVED** [review]
LGTM

File: enterprise/coderd/prebuilds/reconcile.go (hunk 9 above)
Comment (ssncferreira) [reply enterprise/coderd/prebuilds/reconcile.go:149]: We already have `TestReconciliationConcurrency` which covers all the edge cases. It tests via the public getter, but exercises the same logic. But I can remove the getter and test this function internally, that might be cleaner.

Comment (ssncferreira) [reply enterprise/coderd/prebuilds/reconcile.go:149]: Addressed in https://github.com/coder/coder/pull/20908/commits/b2cd71369523127c8c0286267271b71c5c6634b3

**Review by dannykopping — APPROVED** [review]
:shipit:
